### PR TITLE
fix: improve ModelCtor / ModelStatic typing

### PR DESCRIPTION
**Pull request by ephys**
### Pull Request Checklist

_Please make sure to review and check all of these items:_

- [x] Have you added new tests to prevent regressions?
- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you update the typescript typings accordingly (if applicable)?
- [ ] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

### Description Of Change

Alternative PR for https://github.com/sequelize/sequelize/pull/13834

I'm hoping I finally figured out why ModelCtor / ModelStatic are so tricky to use. I tried fixing ModelStatic & Merging both types together instead of having two different ones.

Crossing my fingers for tests to pass.

Edit: this should land in v6

### CODE REVIEW

**Comment by hsource** [comment]
I tested this against our very-large Typescript codebase and it works well! LGTM

File: types/test/type-helpers/deep-writable.ts [hunk 1]
```diff
@@ -6,10 +6,34 @@
...
+  Sequelize,
+  ModelCtor,
+  ModelDefined,
+  ModelStatic,
+} from 'sequelize';
 
-type Builtin = string | number | boolean | bigint | symbol | undefined | null | Function | Date | Error | RegExp;
-type SequelizeBasic = Builtin | Sequelize | Model | ModelCtor<Model> | ModelType | ModelDefined<any, any> | ModelStatic<Model>;
+type Builtin =
+  | string
```
Comment (sdepold) [inline types/test/type-helpers/deep-writable.ts:18]: Is the pipe in the beginning on purpose?

**Review by sdepold — COMMENTED** [review]
I have to admit that I'm still a bit confused by all those types.

Is this correct?:

const User: ModelStatic = sequelize.define('user');
const user:Model = user.findOne();

File: types/test/type-helpers/deep-writable.ts (hunk 1 above)
Comment (ephys) [reply types/test/type-helpers/deep-writable.ts:18]: Yes, it's so all lines look the same no matter whether it's the first element or the last. Good practice for future diffs :)

**Comment by ephys** [comment]
> I have to admit that I'm still a bit confused by all those types.

Your code is correct. ModelStatic represents the class itself, Model represents an instance of the class

`ModelStatic` lets us request the model itself in functions, which can be used to do things like this:

image

image

We used to have 3 different types to represent the class itself (ModelType, ModelCtor, ModelStatic) but this PR makes `ModelStatic` universal and able to replace the other 2 types

**Comment by sdepold** [comment]
>We used to have 3 different types to represent the class itself (ModelType, ModelCtor, ModelStatic) but this PR makes ModelStatic universal and able to replace the other 2 types

Exactly and it drove me half crazy the other day. :praisethesun:

**Comment by sdepold** [comment]
Nice editor theme

**Comment by fncolon** [comment]
What theme is it? @ephys haha

**Comment by ephys** [comment]
@fncolon If I remember right the base was Material Palenight (From Material Theme UI for IntelliJ)

**Comment by fncolon** [comment]
> t

Thank you!

**Pull request merged**